File: NowCoder/牛客周赛/牛客周赛 Round 77/计树.cpp
#include <bits/stdc++.h>
using namespace std;
using LL = long long;
#define endl '\n'

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, u, v, k, x;
    cin >> n;
    vector<vector<int>> g(n);
    for (int i = 0; i < n-1; i++) {
        cin >> u >> v;
        u--; v--;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    cin >> k;
    unordered_set<int> st;
    for (int i = 0; i < k; i++) {
        cin >> x;
        st.insert(x-1);
    }

    vector<LL> cnt(n), ans(n);
    vector<bool> vis(n);
    auto dfs = [&](auto &&dfs, int u, int fa) -> void {
        vis[u] = true;
        cnt[u] = st.count(u) ? 1 : 0;
        
        for (int v : g[u]) {
            if (v == fa || vis[v]) continue;
            dfs(dfs, v, u);
            cnt[u] += cnt[v];
        }
        
        ans[u] = cnt[u] * cnt[u];
        for (int v : g[u]) {
            if (v == fa) continue;
            ans[u] -= cnt[v] * cnt[v];
        }
    };
    dfs(dfs, 0, -1);
    
    for (int i = 0; i < n; i++) {
        cout << ans[i] << (i == n-1 ? '\n' : ' ');
    }
    return 0;
}

File: NowCoder/牛客周赛/牛客周赛 Round 77/计树.py
n = int(input())
g = [[] for _ in range(n)]
for _ in range(n-1):
    u, v = map(lambda x: int(x)-1, input().split())
    g[u].append(v)
    g[v].append(u)
k = int(input())
V = set(map(lambda x: int(x)-1, input().split()))

cnt = [0] * n
ans = [0] * n
vis = [False] * n

# def dfs(u, fa):
#     vis[u] = True
#     cnt[u] = 1 if u in st else 0
     
#     for v in g[u]:
#         if v == fa or vis[v]:
#             continue
#         dfs(v, u)
#         cnt[u] += cnt[v]
     
#     ans[u] = cnt[u] * cnt[u]
#     for v in g[u]:
#         if v == fa:
#             continue
#         ans[u] -= cnt[v] * cnt[v]
# dfs(0, -1)

st = [(0, -1, 0)]  # (u, fa, state)
while st:
    u, fa, state = st.pop()
    if state == 0:
        vis[u] = True
        cnt[u] = 1 if u in V else 0
        st.append((u, fa, 1))
        for v in g[u]:
            if v != fa and not vis[v]:
                st.append((v, u, 0))
    else:
        for v in g[u]:
            if v != fa:
                cnt[u] += cnt[v]
        ans[u] = cnt[u] * cnt[u]
        for v in g[u]:
            if v != fa:
                ans[u] -= cnt[v] * cnt[v]
print(*ans)